tasmota/tasmota_xsns_sensor/xsns_33_qmc5883l.ino

```diff
@@ -0,0 +1,385 @@
+
+  if (I2cValidReadS16(&QMC5883L.MX, QMC5883L.i2c_address, QMC5883L_X_LSB) == false)  return; // read error, select LSB register
+  if (I2cValidReadS16(&QMC5883L.MY, QMC5883L.i2c_address, QMC5883L_Y_LSB) == false)  return; // read error, select LSB register
+  if (I2cValidReadS16(&QMC5883L.MZ, QMC5883L.i2c_address, QMC5883L_Z_LSB) == false)  return; // read error, select LSB register
+
+  int16_t x = QMC5883L.MX;
+  int16_t y = QMC5883L.MY;
+
+  // calculate azimut, heading
+  if(x < QMC5883L.xlow) QMC5883L.xlow = x;
+  if(x > QMC5883L.xhigh) QMC5883L.xhigh = x;
+  if(y < QMC5883L.ylow) QMC5883L.ylow = y;
+  if(y > QMC5883L.yhigh) QMC5883L.yhigh = y;
+
+  /* Bail out if not enough data is available. */
+  if( QMC5883L.xlow == QMC5883L.xhigh || QMC5883L.ylow == QMC5883L.yhigh ) return;
+  /* Recenter the measurement by subtracting the average */
+  x -= (QMC5883L.xhigh + QMC5883L.xlow) / 2;
+  y -= (QMC5883L.yhigh + QMC5883L.ylow) / 2;
+  /* Rescale the measurement to the range observed. */
+  float fx = (float) x / (QMC5883L.xhigh - QMC5883L.xlow);
+  float fy = (float) y / (QMC5883L.yhigh - QMC5883L.ylow);
+
+  x = -atan2(fy, fx) * 180.0 / M_PI;
+  x += 180;  // no negative numbers
+  QMC5883L.HG = x;
+
+  // calculate scalar magnetic induction
+  QMC5883L.scalar = sqrt((QMC5883L.MX * QMC5883L.MX) + (QMC5883L.MY * QMC5883L.MY) + (QMC5883L.MZ * QMC5883L.MZ));
+ 
+   // get temperature
+  //if (readRegister(QMC5883L_TEMP_LSB, 2) != 2)   return; // read error
+  if (I2cValidReadS16(&QMC5883L.temp, QMC5883L.i2c_address, QMC5883L_TEMP_LSB) == false)  return; // read error, select LSB register
+  //t = Wire.read() | (Wire.read() << 8);
+  QMC5883L.temp = (QMC5883L.temp / 100) + USE_QMC5883L_Temp;
+}
+
+/*********************************************************************************************\
+ * Presentation
+\*********************************************************************************************/
+
+#ifdef USE_WEBSERVER
+const char HTTP_SNS_QMC5883L[] PROGMEM =
+    "{s}QMC5883L " D_MX "{m}%d " D_UNIT_MICROTESLA "{e}"                        // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+    "{s}QMC5883L " D_MY "{m}%d " D_UNIT_MICROTESLA "{e}"                        // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+    "{s}QMC5883L " D_MZ "{m}%d " D_UNIT_MICROTESLA "{e}"                        // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+    "{s}QMC5883L " D_MAGNETICFLD "{m}%d " D_UNIT_MICROTESLA "{e}"               // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+    "{s}QMC5883L " D_HG "{m}%d " D_UNIT_DEGREE "{e}"                            // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+    "{s}QMC5883L " D_TEMPERATURE "{m}%d " D_UNIT_DEGREE D_UNIT_CELSIUS "{e}"; // {s} = <tr><th>, {m} = </th><td>, {e} = </td></tr>
+
+const char HTTP_SNS_QMC5883L_ERROR[] PROGMEM =
+    "{s}QMC5883L {m} %s {e}";
+#endif
+
+
+void QMC5883L_Show(uint8_t json)
+{
+  if (json) 
+  {
+    if (QMC5883L.ready != true)
```

Comment:
You don't need this test as you should never come here is the init hasnt been successfull